rasterio/__init__.py

```diff
@@ -254,8 +254,11 @@ def open(
             or isinstance(fp, (os.PathLike, MemoryFile, FilePath))
         ):
             raise TypeError(f"invalid path or file: {fp!r}")
-    if mode and not isinstance(mode, str):
-        raise TypeError(f"invalid mode: {mode!r}")
+    if mode:
+        if not isinstance(mode, str):
```

Comment:
do we really need to test for the type? 

could we just do `if not mode.startswith(("r", "w")):` ? 

doing this change will mean that the error which might be raised will be something else than `TypeError`
```python

In [5]: a = []

In [6]: a.startswith(("r", "w"))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-6af35dd62cb3> in <cell line: 1>()
----> 1 a.startswith(("r", "w"))

AttributeError: 'list' object has no attribute 'startswith'
```
🤷